src/java.base/share/classes/jdk/internal/lang/stable/StableValueImpl.java

```diff
@@ -0,0 +1,419 @@
+        if (isSet() || isError()) {
+            return false;
+        }
+        final var m = acquireMutex();
+        if (isMutexNotNeeded(m)) {
+            return false;
+        }
+        synchronized (m) {
+            if (isSet() || isError()) {
+                return false;
+            }
+            setValue(value);
+            return true;
+        }
+    }
+
+    @ForceInline
+    @Override
+    public V computeIfUnset(Supplier<? extends V> supplier) {
+        // Optimistically try plain semantics first
+        final V v = value;
+        if (v != null) {
+            // If we happen to see a non-null value under
+            // plain semantics, we know a value is set.
+            return v;
+        }
+        if (state == NULL) {
+            // If we happen to see a state value of NULL under
+            // plain semantics, we know a value is set to `null`.
+            return null;
+        }
+        // Now, fall back to volatile semantics.
+        return computeIfUnsetVolatile(supplier);
+    }
+
+    @DontInline // Slow-path taken at most once per thread if set
+    private V computeIfUnsetVolatile(Supplier<? extends V> supplier) {
+        // This is intentionally an old switch statement as it generates
+        // more compact byte code.
+        switch (stateVolatile()) {
+            case UNSET:    { return computeIfUnsetVolatile0(supplier); }
+            case NON_NULL: { return valueVolatile(); }
+            case NULL:     { return null; }
+            case ERROR:    { throw StableUtil.error(this); }
+        }
+        throw shouldNotReachHere();
+    }
+
+    private V computeIfUnsetVolatile0(Supplier<? extends V> supplier) {
+        final var m = acquireMutex();
+        if (isMutexNotNeeded(m)) {
+            return orThrow();
+        }
+        synchronized (m) {
+            // A value is already set
+            if (state != UNSET) {
+                return orThrow();
+            }
+            // A value is not set
+            if (supplying) {
+                throw stackOverflow(supplier, null);
+            }
+            try {
+                supplying = true;
+                try {
+                    V newValue = supplier.get();
+                    setValue(newValue);
+                    return newValue;
+                } catch (Throwable t) {
+                    putState(ERROR);
+                    putMutex(t.getClass());
+                    throw t;
+                }
+            } finally {
+                supplying = false;
+            }
+        }
+    }
+
+    @ForceInline
+    public V computeIfUnset(int index, IntFunction<? extends V> mapper) {
+        return computeIfUnsetShared(mapper, index);
+    }
+
+    @ForceInline
+    public <K> V computeIfUnset(K key, Function<? super K, ? extends V> mapper) {
+        return computeIfUnsetShared(mapper, key);
+    }
+
+    @ForceInline
+    private <K> V computeIfUnsetShared(Object provider, K key) {
```

Comment:
Can we let suppliers share this path too, with a null key? I see this path supports suppliers but supplier code path doesn't call this path.

Reply:
Good suggestion!